Domain Management i1: show indicators when bulk jobs are happening

mpkelly opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/3419

## Proposed Changes

It's a work in progress but is starting to come together:

* Support polling for bulk status updates and showing jobs in progress
    - Checks on load and decides whether to keep polling or not
    - Shows spinner on domains with pending updates
    - Shows global status of all running bulk jobs  
    - calls API to delete old jobs when popover is closed. This will clear the top-level notification the next time `/GET domains/bulk-actions` is called if no other jobs are running. 

**The global popup shows domain-level status updates. This is a rabbit hole, I feel.  Maybe we just want to show that a job is in progress and then report any errors if something was wrong?**

**Global level**:

<img width="982" alt="Screenshot 2566-09-05 at 12 06 29" src="https://github.com/Automattic/wp-calypso/assets/6851384/e83ad927-8bb1-4e42-bafb-b4e877fd236d">

![image](https://github.com/Automattic/wp-calypso/assets/6851384/f5c084df-f1ae-4e47-bf42-07afff038f51)


**Domain level:**
![image](https://github.com/Automattic/wp-calypso/assets/6851384/23b03334-398e-4917-b83c-a573536fe66a)



## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply D120555-code your sandbox, so deleting completed jobs will work
* Checkout this branch
* Navigate to `domains/manage?flags=domains/management`
* Select one or more domains and submit a bulk auto-renew update
* Confirm that you see the indicators above on the domain level and in the header
* Open your browser's Network tools and confirm that `domains/bulk-actions` is being called every six seconds
* Make sure your sandbox is set up for the new async jobs as per "configure sandbox" here: D119397-code 
* Go to your sandbox and run `cat /tmp/php-errors` and get the async command. 
* Run the command, e.g. `async-jobs/tester.php 87169417713 wpj_` 
    -  Watch the domain(s) you updated and the spinner should clear. 
    - At global level, the bulk job should go from orange to green status
    - The API calls to `domains/bulk-actions` should stop 
* For additional tests, you can try modifying endpoint `GET /domains/bulk-actions` in `wp-content/rest-api-plugins/endpoints/domains-bulk-actions.php` to return different payloads, or just modify `packages/data-stores/src/queries/use-bulk-domain-update-status-query.ts` in the client. 


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

mpkelly committed:
Merge with trunk

mpkelly committed:
Merge with trunk

mpkelly committed:
Show more info about running jobs

mpkelly committed:
Uses Notice instead of custom UI.

mpkelly committed:
Translate text

p-jackson commented on packages/data-stores/src/queries/use-bulk-domain-update-status-query.ts:
```diff
@@ -0,0 +1,106 @@
+export enum BulkDomainUpdateStatusRetryInterval {
+	Active = 6000,
+	Disabled = -1,
+}
```
We generally try to avoid enums p7H4VZ-3sK-p2. An alternative might be for `useBulkDomainUpdateStatusQuery()` to accept a `pollingEnabled` boolean as its parameter? Or to keep the interval generic maybe it accepts `number|null` and the calling code can use its own constant rather than importing the constant from this module.

mpkelly replied:
I have switched this to a constant object.

p-jackson commented on packages/domains-table/src/status-popover/index.tsx:
```diff
@@ -8,9 +8,16 @@ import './style.scss';
 type StatusPopoverProps = {
 	children: React.ReactNode;
 	className?: string;
+	target?: React.ReactNode;
```
Maybe rename this prop to `icon`? It took me a while to realise what was meant by "target".

mpkelly replied:
It's not always an icon, so I called it `popoverTargetElement` instead.

p-jackson commented on packages/domains-table/src/use-domains-table.ts:
```diff
@@ -6,7 +6,7 @@ import type { UseQueryOptions } from '@tanstack/react-query';
 
 export function useDomainsTable( queryOptions: UseQueryOptions< AllDomainsQueryFnData > = {} ) {
-	const { capabilities, sites } = useSelector( ( state: any ) => ( {
+	const { capabilities, sites } = useSelector( ( state ) => ( {
```
The removal of the `any` annotation is giving me errors. We're being a bit naughty here and accessing the Calypso redux store from within a package, which isn't guaranteed to work. It's only temporary because we're going to change the endpoint we fetch from so that it doesn't return us domains we can't manage in the first place. The fact we can't get the type quite right is a symptom of the fact we're being naughty. Technically we should use the `IAppState` type, but that's defined in `client/state/types` which we can't import.

Is it possible to go back to the `any` for now?

mpkelly committed:
Use a default value for pendingUpdates

p-jackson commented on packages/domains-table/src/use-domain-bulk-update-status.ts:
```diff
@@ -0,0 +1,50 @@
+		// be enough
+		setTimeout( () => {
+			setStatusUpdateInterval( BulkDomainUpdateStatusRetryInterval.Active );
+			queryClient.invalidateQueries( { queryKey: [ 'domains', 'bulk-actions' ] } );
```
We often export a function like `getBulkDomainUpdateStatusQueryKey()` to ensure the same query key is used everywhere e.g. https://github.com/Automattic/wp-calypso/blob/be1d326e40784a89b5594dd66b84297d831a6514/packages/data-stores/src/queries/use-site-query.ts#L23

p-jackson replied:
Although I see from the todo this is just temporary ...

mpkelly replied:
I moved it into a function as we don't know how long this "temp" code will stick around.

p-jackson reviewed (approved):
I had a bit of trouble getting my test environment set up, but turns out that's because I had already sandboxed the async jobs when I hadn't expected to. So I ended up with some bulk actions which I couldn't get rid of, even by calling the `DELETE` endpoint manually since that endpoint only cleans up the completed tasks.

It got me thinking, should we have a failsafe in our async jobs? In case the jobs get stuck in a running state for weeks. Should we have a fail safe that they cancel after some amount of time.

There's some discussion about the notice UI here: p1693905676373519/1693623343.356829-slack-C04H4NY6STW
We don't have to implement it all in this PR, but is there some way to convert the current notice component into the hovering one? So that it floats at the top right of the page rather than taking up vertical space in the document?

Left some notes, but everything worked as expected 👍

p-jackson commented:
@mpkelly oops left out one bit of feedback. When I dismiss the success notice I expect the UI to disappear immediately, however it's currently waiting until the `DELETE` request is complete. I think we can optimistically dismiss the notice.

mpkelly commented:
> It got me thinking, should we have a failsafe in our async jobs? In case the jobs get stuck in a running state for weeks. Should we have a fail safe that they cancel after some amount of time.

Sounds good to me. So this would be something in the domain row popup? I need to include the timestamp in that popup too.

mpkelly committed:
use any type to fix errors.

mpkelly committed:
Use more descriptive name for popover target element

mpkelly committed:
Use query key function

mpkelly committed:
Notice should disappear instantly.

p-jackson commented:
> Sounds good to me. So this would be something in the domain row popup?

I wasn't thinking something the user could control. What I'm wondering is if we should have some sort of fail safe on the backend. To stop us from filling up the jobs table and hogging all the job servers if there's a bug.

Maybe we should just send a message to #dotcom-build-errors if a job has been running for more than a day.

mpkelly committed:
Enums considered harmful :p

mpkelly commented:
@p-jackson, I have addressed those review comments, except for the failsafe. 

>Maybe we should just send a message to #dotcom-build-errors if a job has been running for more than a day.

I wonder what would keep a job running for over a day? Could each update really be so slow? If there is an error it should already get handled.  The current backend code will remove old jobs from the attribute, but this clean-up only happens when a new job comes in. 

I wonder if the Systems team already monitor jobs that haven't stopped running, as these will affect overall throughput.

mpkelly committed:
Show time bulk action was

p-jackson commented:
> I wonder what would keep a job running for over a day?

I'm not sure. Just a coding bug I guess.
I'm trying to figure out whether the situation I ran into (where I had a spinner that I couldn't get rid of) could happen to a user. Not because of an action taking a long time, just because of a bug.

On reflection, perhaps orphaned jobs isn't the issue I was having. It's that the user attribute was out of sync with reality. The user attribute said that it had kicked off a job, but because the job was sandboxed it had never really kicked off. That's not exactly something that can happen to a user, but I'm just trying to think, is there a way that could have been automatically fixed up so that a user won't see a similar thing if there's a bug in our code?

p-jackson commented:
@mpkelly looks like the build error is coming from importing the `calypso/components/notice` file. The problem is importing js into a tsx file I think.

You can silence it by adding this to `packages/domains-table/src/type-declarations.d.ts`

```ts
declare module 'calypso/components/notice';
```

mpkelly commented:
> @mpkelly looks like the build error is coming from importing the calypso/components/notice file. The problem is importing js into a tsx file I think.

There are some types for this `Notice` component. They aren't found because we are in a `package` and the `Notice` is part of the project that imports the packages - i.e. the dependency relationship is the opposite direction to how I am using it (now I understand why there is a lint rule). I am going to try the `declare module 'calypso/components/notice';` as a quick fix.

mpkelly committed:
Fix build errors.

mpkelly commented:
The timestamp is now shown on the update:

![image](https://github.com/Automattic/wp-calypso/assets/6851384/df9d2691-58df-4e9c-b658-a27cf6ef4f1b)

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8868848

Thank you @mpkelly for including a screenshot in the description! This is really helpful for our translators.
